2.0-proposal/RELATIONS.md

```diff
@@ -0,0 +1,248 @@
+to which one or more remote applications can be connected (related).  A layer
+which provides an interface implementation is a normal layer, but has to
+conform to a certain set of requirements.  The layer must provide one or two
+files with specific paths, containing classes that implement both the provides
+and requires side of an interface, or the single peer interface impelmentation.
+
+The files should be under the `reactive/relations/` directory, in a
+sub-directory named after the interface name (with dashes changed to
+underscores), and in files named one of `provides.py`, `requires.py`,
+or `peers.py`.  Thus, the layer implementing the `apache-website` interface
+would need to provide the following files:
+
+  * `reactive/relations/apache_website/provides.py`
+  * `reactive/relations/apache_website/requires.py`
+
+The class in each of these files must be a subclass of
+`charms.reactive.SimpleRelation`.
+
+## Documenting Interface Layers
+
+The classes provided by an interface layer are expected to use docstrings to
+well document their API.  There is a `charm interface-doc` command that will
+convert the docstrings into Markdown and inject them into the `README.md` file
+between special comments of the form: &lt;!--interface-doc--&gt; and
+&lt;!--/interface-doc--&gt;.  This way, all interface layers have consistent
+and comprehensive documentation.
+
+# Accessing Relation Instances
+
+An instance of the appropriate interface layer class will be created and
+added to the global reactive context variable, where it can be accessed
+by the name of the relation (with dashes changed to underscores).  For example,
+if a charm had the following in its `metadata.yaml`:
+
+```yaml
+provides:
+  apache-website:
+    interface: apache-website
+requires:
+  db:
+    interface: mysql
+peers:
+  peers:
+    interface: myapp
+```
+
+Then the following context attributes would be available:
+
+```python
+from charms.reactive import context
+
+assert context.relations.apache_website
+assert context.relations.db
+assert context.relations.peers
+```
+
+Note that the `apache_website` instance would be of the class that was found in
+the `reactive/relations/apache_website/provides.py` file, the `db` instance
+would be from `reactive/relations/mysql/requires.py`, and the `peers` instance
+would be from `reactive/relations/myapp/peers.py`, in their respective layers.
+
+Also note that those instances would be available even if their relation is not
+related to any remote application.
+
+# The SimpleRelation API
+
+The `SimpleRelation` base class will provide a couple of automatically managed
+reactive flags, as well as providing collections for introspecting the state of
+the relation, including related applications and units, relation data provided
+by those units, and for associating reactive flags with particular related
+applications or units and filtering related applications and units by those
+flags.
+
+## Managed Flags
```

Comment:
@johnsca Will these flags be available to layers, or only to the specific interface layer that manages the relationship?